File: ios/Objective-C/GeneralSettingsObjC/GeneralSettingsObjC/Modules/BarcodeSettings/BarcodeSettingsViewController.m
//
//  BarcodeSettingsViewController.m
//  GeneralSettings
//
//  Created by dynamsoft on 2021/11/19.
//

#import "BarcodeSettingsViewController.h"
#import "BasicSwitchTableViewCell.h"
#import "BarcodeFormatDetailViewController.h"

// The barcode formats and barcode count settings are common for barcode reader.
// Specifying the barcode formats will help you improve the performance of the barcode reader.
// The less barcode count, the higher processing speed.
// For unknown usage scenarios, set the barcode count to 0. The barcode reader will try to decode at least one barcode.

@interface BarcodeSettingsViewController ()<UITableViewDelegate, UITableViewDataSource>
{
    NSArray *barcodeSettingDataArray;
    
    /**
     Record DBR switch state dic.
     */
    NSMutableDictionary *recordDBRSwitchStateDic;
    
    NSInteger expectedCount;
    NSInteger minimumResultConfidence;
    NSInteger duplicateForgetTime;
    NSInteger minimumDecodeInterval;
    
    
}
@property (nonatomic, strong) UITableView *barcodeSettingTableView;

@end

@implementation BarcodeSettingsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.title = @"Barcode Settings";
    
    [self handleData];
    [self setupUI];
}

- (void)handleData {
   
    recordDBRSwitchStateDic = [NSMutableDictionary dictionary];
    
    // Array.
    if ([GeneralSettingsHandle setting].barcodeSettings.duplicateFliterIsOpen) {
        barcodeSettingDataArray = @[[GeneralSettingsHandle setting].barcodeSettings.barcodeFormatStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.expectedCountStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.continuousScanStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.minimumResultConfidenceStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.duplicateForgetTimeStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.minimumDecodeIntervalStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr
        ];
    } else {
        barcodeSettingDataArray = @[[GeneralSettingsHandle setting].barcodeSettings.barcodeFormatStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.expectedCountStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.continuousScanStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.minimumResultConfidenceStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.minimumDecodeIntervalStr,
                                    [GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr
        ];
    }
    
    // Switch state dic.
    if ([GeneralSettingsHandle setting].barcodeSettings.continuousScanIsOpen == YES) {
        [recordDBRSwitchStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].barcodeSettings.continuousScanStr];
    } else {
        [recordDBRSwitchStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].barcodeSettings.continuousScanStr];
    }
    
    if ([GeneralSettingsHandle setting].barcodeSettings.resultVerificationIsOpen == YES) {
        [recordDBRSwitchStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr];
    } else {
        [recordDBRSwitchStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr];
    }
    
    if ([GeneralSettingsHandle setting].barcodeSettings.duplicateFliterIsOpen == YES) {
        [recordDBRSwitchStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr];
    } else {
        [recordDBRSwitchStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr];
    }
    
    if ([GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesIsOpen == YES) {
        [recordDBRSwitchStateDic setValue:@(1) forKey:[GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr];
    } else {
        [recordDBRSwitchStateDic setValue:@(0) forKey:[GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr];
    }
    
    expectedCount = [GeneralSettingsHandle setting].barcodeSettings.expectedCount;
    minimumResultConfidence = [GeneralSettingsHandle setting].barcodeSettings.minimumResultConfidence;
    duplicateForgetTime = [GeneralSettingsHandle setting].barcodeSettings.duplicateForgetTime;
    minimumDecodeInterval = [GeneralSettingsHandle setting].barcodeSettings.minimumDecodeInterval;
}

- (void)setupUI
{
    self.barcodeSettingTableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.barcodeSettingTableView.backgroundColor = [UIColor whiteColor];
    self.barcodeSettingTableView.delegate = self;
    self.barcodeSettingTableView.dataSource = self;
    self.barcodeSettingTableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    [self.view addSubview:self.barcodeSettingTableView];
}


// MARK: - UITableViewDelegate
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return barcodeSettingDataArray.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [BasicTableViewCell cellHeight];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *titleString = barcodeSettingDataArray[indexPath.row];
    
    weakSelfs(self)
    if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.barcodeFormatStr]) {
        static NSString *identifier = @"basicCellIdentifier";
        BasicTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        
        [cell updateUIWithTitle:titleString];
        return cell;
        
    } else if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.expectedCountStr]) {
        static NSString *identifier = @"basicTextCellIdentifier";
        BasicTextTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicTextTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.questionBlock = ^{
            [weakSelf handleDBRExplainWithIndexPath:indexPath settingString:titleString];
        };
        
        cell.inputTFValueChangedBlock = ^(NSInteger numValue) {
            [weakSelf handleExpectedCount:numValue];
        };
        
        [cell setInputCountTFMaxValueWithNum:kExpectedCountMaxValue];
        [cell setDefaultValue:expectedCount];
        [cell updateUIWithTitle:titleString value:expectedCount];
        return cell;
        
    } else if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.minimumResultConfidenceStr]) {
        static NSString *identifier = @"basicTextCellIdentifier";
        BasicTextTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicTextTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        
        cell.inputTFValueChangedBlock = ^(NSInteger numValue) {
            [weakSelf handleMinimumResultConfidence:numValue];
        };
        
        [cell setInputCountTFMaxValueWithNum:kMinimumResultConfidenceMaxValue];
        [cell setQuestionButtonIsHidden:YES];
        [cell setDefaultValue:minimumResultConfidence];
        [cell updateUIWithTitle:titleString value:minimumResultConfidence];
        return cell;
        
    } else if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateForgetTimeStr]) {
        static NSString *identifier = @"basicTextCellIdentifier";
        BasicTextTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicTextTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.questionBlock = ^{
            [weakSelf handleDBRExplainWithIndexPath:indexPath settingString:titleString];
        };
        
        cell.inputTFValueChangedBlock = ^(NSInteger numValue) {
            [weakSelf handleDuplicateForgetTime:numValue];
        };
        
        [cell setInputCountTFMaxValueWithNum:kDuplicateForgetTimeMaxValue];
        [cell setTitleOffset:20.0];
        [cell setDefaultValue:duplicateForgetTime];
        [cell updateUIWithTitle:titleString value:duplicateForgetTime];
        return cell;
        
    } else if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.minimumDecodeIntervalStr]) {
        static NSString *identifier = @"basicTextCellIdentifier";
        BasicTextTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicTextTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.questionBlock = ^{
            [weakSelf handleDBRExplainWithIndexPath:indexPath settingString:titleString];
        };
        
        cell.inputTFValueChangedBlock = ^(NSInteger numValue) {
            [weakSelf handleMinimumDecodeInterval:numValue];
        };
        
        [cell setInputCountTFMaxValueWithNum:kMinimumDecodeIntervalMaxValue];
        [cell setDefaultValue:minimumDecodeInterval];
        [cell updateUIWithTitle:titleString value:minimumDecodeInterval];
        return cell;
        
    } else if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.continuousScanStr] ||
               [titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr] ||
               [titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr] ||
               [titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr]) {
        NSInteger switchSelectState = [[recordDBRSwitchStateDic valueForKey:titleString] integerValue];
        static NSString *identifier = @"basicSwitchCellIdentifier";
        BasicSwitchTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        if (!cell) {
            cell = [[BasicSwitchTableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
        }
        
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr] ||
            [titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr]) {
            cell.questionButton.hidden = NO;
        } else {
            cell.questionButton.hidden = YES;
        }
        cell.questionBlock = ^{
            [weakSelf handleDBRExplainWithIndexPath:indexPath settingString:titleString];
        };
       
        cell.switchChangedBlock = ^(BOOL isOn) {
            [weakSelf handleDBRSettingSwitchWithIndexPath:indexPath settingString:titleString andSwitchState:isOn];
        };
 
        [cell updateUIWithTitle:titleString withSwitchState:switchSelectState];
        return cell;
    }
    
    return [UITableViewCell new];
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    NSString *titleString = barcodeSettingDataArray[indexPath.row];
    if ([titleString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.barcodeFormatStr]) {
       
        BarcodeFormatDetailViewController *barcodeFormatDetailVC = [[BarcodeFormatDetailViewController alloc] init];
        [self.navigationController pushViewController:barcodeFormatDetailVC animated:YES];
    }
  
}

// MARK: - Handle value changed
- (void)handleExpectedCount:(NSInteger)numValue {
    BarcodeSettings barcodeSettings = [GeneralSettingsHandle setting].barcodeSettings;
    barcodeSettings.expectedCount = numValue;
    [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
    
    [GeneralSettingsHandle setting].ipublicRuntimeSettings.expectedBarcodesCount = numValue;
    [[GeneralSettingsHandle setting] updateIpublicRuntimeSettings];
    
    [self handleData];
    [self.barcodeSettingTableView reloadData];
}

- (void)handleMinimumResultConfidence:(NSInteger)numValue {
    BarcodeSettings barcodeSettings = [GeneralSettingsHandle setting].barcodeSettings;
    barcodeSettings.minimumResultConfidence = numValue;
    [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
    
    [GeneralSettingsHandle setting].ipublicRuntimeSettings.minResultConfidence = numValue;
    [[GeneralSettingsHandle setting] updateIpublicRuntimeSettings];
    
    [self handleData];
    [self.barcodeSettingTableView reloadData];
}

- (void)handleDuplicateForgetTime:(NSInteger)numValue {
    BarcodeSettings barcodeSettings = [GeneralSettingsHandle setting].barcodeSettings;
    barcodeSettings.duplicateForgetTime = numValue;
    [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
    
    [[GeneralSettingsHandle setting].barcodeReader setDuplicateForgetTime:numValue];
    
    [self handleData];
    [self.barcodeSettingTableView reloadData];
}

- (void)handleMinimumDecodeInterval:(NSInteger)numValue {
    BarcodeSettings barcodeSettings = [GeneralSettingsHandle setting].barcodeSettings;
    barcodeSettings.minimumDecodeInterval = numValue;
    [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
    
    [[GeneralSettingsHandle setting].barcodeReader setMinImageReadingInterval:numValue];
    
    [self handleData];
    [self.barcodeSettingTableView reloadData];
}

// MARK: - Handle switch changed
- (void)handleDBRSettingSwitchWithIndexPath:(NSIndexPath *)indexPath settingString:(NSString *)dbrSettingString andSwitchState:(BOOL)isOn {
    
    BarcodeSettings barcodeSettings = [GeneralSettingsHandle setting].barcodeSettings;
    
    if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.continuousScanStr]) {
        barcodeSettings.continuousScanIsOpen = isOn;
        [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr]) {
        barcodeSettings.resultVerificationIsOpen = isOn;
        [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
        
        [GeneralSettingsHandle setting].barcodeReader.enableResultVerification = isOn;
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr]) {
        barcodeSettings.duplicateFliterIsOpen = isOn;
        [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
        
        [GeneralSettingsHandle setting].barcodeReader.enableDuplicateFilter = isOn;
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.decodeInvertedBarcodesStr]) {
        barcodeSettings.decodeInvertedBarcodesIsOpen = isOn;
        [GeneralSettingsHandle setting].barcodeSettings = barcodeSettings;
        
        NSArray *grayscaleTransformationModes = @[];
        if (isOn) {
            grayscaleTransformationModes = @[@(EnumGrayscaleTransformationModeOriginal),
                                             @(EnumGrayscaleTransformationModeInverted),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip)
            ];
        } else {
            grayscaleTransformationModes = @[@(EnumGrayscaleTransformationModeOriginal),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip),
                                             @(EnumGrayscaleTransformationModeSkip)
            ];
        }
        [GeneralSettingsHandle setting].ipublicRuntimeSettings.furtherModes.grayscaleTransformationModes = grayscaleTransformationModes;
        [[GeneralSettingsHandle setting] updateIpublicRuntimeSettings];
    }
    
    if (isOn == YES) {
        [recordDBRSwitchStateDic setValue:@(1) forKey:dbrSettingString];
    } else {
        [recordDBRSwitchStateDic setValue:@(0) forKey:dbrSettingString];
    }
    
    [self handleData];
    [self.barcodeSettingTableView reloadData];
}

- (void)handleDBRExplainWithIndexPath:(NSIndexPath *)indexPath settingString:(NSString *)dbrSettingString {
    if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.expectedCountStr]) {
        [[ToolsHandle toolManger] addAlertViewWithTitle:dbrSettingString Content:expectedCountExplain actionTitle:nil ToView:self completion:nil];
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateForgetTimeStr]) {
        [[ToolsHandle toolManger] addAlertViewWithTitle:dbrSettingString Content:duplicateForgetTimeExplain actionTitle:nil ToView:self completion:nil];
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.minimumDecodeIntervalStr]) {
        [[ToolsHandle toolManger] addAlertViewWithTitle:dbrSettingString Content:minimumDecodeIntervalExplain actionTitle:nil ToView:self completion:nil];
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.resultVerificationStr]) {
        [[ToolsHandle toolManger] addAlertViewWithTitle:dbrSettingString Content:resultVerificationExplain actionTitle:nil ToView:self completion:nil];
    } else if ([dbrSettingString isEqualToString:[GeneralSettingsHandle setting].barcodeSettings.duplicateFliterStr]) {
        [[ToolsHandle toolManger] addAlertViewWithTitle:dbrSettingString Content:duplicateFilterExplain actionTitle:nil ToView:self completion:nil];
    }
}

@end

File: ios/Swift/GeneralSettingsSwift/GeneralSettingsSwift/ViewControllers/BarcodeSettings/BarcodeSettingsViewController.swift
//
//  BarcodeSettingsViewController.swift
//  GeneralSettingsSwift
//
//  Copyright © Dynamsoft. All rights reserved.
//

import UIKit

class BarcodeSettingsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    private var dbrSettingsDataArray: [String] = []
    
    /// Record DBR switch state dic.
    private var recordDBRSwitchStateDic: [String : Bool] = [:]
    
    private var expectedCount = 0
    
    private var minimumResultConfidence = 0
    
    private var duplicateForgetTime = 0
    
    private var minimumDecodeInterval = 0
    
    private lazy var dbrSettingsTableView: UITableView = {
        let tableView = UITableView.init(frame: self.view.bounds, style: .plain)
        tableView.backgroundColor = .white
        tableView.delegate = self
        tableView.dataSource = self
        tableView.separatorStyle = .none
        return tableView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        self.view.backgroundColor = .white
        self.title = "Barcode Settings"
        
        handleData()
        setupUI()
    }
    
    private func handleData() -> Void {
        dbrSettingsDataArray.removeAll()
        recordDBRSwitchStateDic.removeAll()
        
        if GeneralSettings.shared.barcodeSettings.duplicateFliterIsOpen == true {
            dbrSettingsDataArray = [GeneralSettings.shared.barcodeSettings.barcodeFormatStr,
                                    GeneralSettings.shared.barcodeSettings.expectedCountStr,
                                    GeneralSettings.shared.barcodeSettings.continuousScanStr,
                                    GeneralSettings.shared.barcodeSettings.minimumResultConfidenceStr,
                                    GeneralSettings.shared.barcodeSettings.resultVerificationStr,
                                    GeneralSettings.shared.barcodeSettings.duplicateFliterStr,
                                    GeneralSettings.shared.barcodeSettings.duplicateForgetTimeStr,
                                    GeneralSettings.shared.barcodeSettings.minimumDecodeIntervalStr,
                                    GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr
            ]
        } else {
            dbrSettingsDataArray = [GeneralSettings.shared.barcodeSettings.barcodeFormatStr,
                                    GeneralSettings.shared.barcodeSettings.expectedCountStr,
                                    GeneralSettings.shared.barcodeSettings.continuousScanStr,
                                    GeneralSettings.shared.barcodeSettings.minimumResultConfidenceStr,
                                    GeneralSettings.shared.barcodeSettings.resultVerificationStr,
                                    GeneralSettings.shared.barcodeSettings.duplicateFliterStr,
                                    GeneralSettings.shared.barcodeSettings.minimumDecodeIntervalStr,
                                    GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr
            ]
        }
        
        // Switch state dic.
        if GeneralSettings.shared.barcodeSettings.continuousScanIsOpen == true {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.continuousScanStr] = true
        } else {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.continuousScanStr] = false
        }
        
        if GeneralSettings.shared.barcodeSettings.resultVerificationIsOpen == true {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.resultVerificationStr] = true
        } else {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.resultVerificationStr] = false
        }
        
        if GeneralSettings.shared.barcodeSettings.duplicateFliterIsOpen == true {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.duplicateFliterStr] = true
        } else {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.duplicateFliterStr] = false
        }
        
        if GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesIsOpen == true {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr] = true
        } else {
            recordDBRSwitchStateDic[GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr] = false
        }
        
        expectedCount = GeneralSettings.shared.barcodeSettings.expectedCount
        minimumResultConfidence = GeneralSettings.shared.barcodeSettings.minimumResultConfidence
        duplicateForgetTime = GeneralSettings.shared.barcodeSettings.duplicateForgetTime
        minimumDecodeInterval = GeneralSettings.shared.barcodeSettings.minimumDecodeInterval
    }
    
    private func setupUI() -> Void {
        self.view.addSubview(dbrSettingsTableView)
    }

    // MARK: - UITableViewDelegate
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return BasicTableViewCell.cellHeight()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.dbrSettingsDataArray.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let titleTag = self.dbrSettingsDataArray[indexPath.row]
        
        if titleTag == GeneralSettings.shared.barcodeSettings.barcodeFormatStr {
            
            let identifier = "BasicTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicTableViewCell
            if cell == nil {
                cell = BasicTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
            cell?.accessoryType = .disclosureIndicator
            cell?.updateUI(with: titleTag)
            return cell!
        } else if titleTag == GeneralSettings.shared.barcodeSettings.expectedCountStr {
            
            let identifier = "BasicTextTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicTextTableViewCell
            if cell == nil {
                cell = BasicTextTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
           
            cell?.questionCompletion = {
                [unowned self] in
                self.handleDBRExplain(with: indexPath, settingString: titleTag)
            }
            
            cell?.inputTFValueChangedCompletion = {
                [unowned self] value in
                self.handleInputValueChanged(with: titleTag, numValue: value)
            }
            
            cell?.setInputCountTFMaxValue(with: kExpectedCountMaxValue)
            cell?.defaultValue = expectedCount
            cell?.updateUI(with: titleTag, valueNum: expectedCount)
            return cell!
        } else if titleTag == GeneralSettings.shared.barcodeSettings.minimumResultConfidenceStr {
            
            let identifier = "BasicTextTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicTextTableViewCell
            if cell == nil {
                cell = BasicTextTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
            
            cell?.inputTFValueChangedCompletion = {
                [unowned self] value in
                self.handleInputValueChanged(with: titleTag, numValue: value)
            }
            
            cell?.setInputCountTFMaxValue(with: kMinimumResultConfidenceMaxValue)
            cell?.questionButtonIsHidden = true
            cell?.defaultValue = minimumResultConfidence
            cell?.updateUI(with: titleTag, valueNum: minimumResultConfidence)
            return cell!
        }  else if titleTag == GeneralSettings.shared.barcodeSettings.duplicateForgetTimeStr {
            
            let identifier = "BasicTextTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicTextTableViewCell
            if cell == nil {
                cell = BasicTextTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
            
            cell?.questionCompletion = {
                [unowned self] in
                self.handleDBRExplain(with: indexPath, settingString: titleTag)
            }
            
            cell?.inputTFValueChangedCompletion = {
                [unowned self] value in
                self.handleInputValueChanged(with: titleTag, numValue: value)
            }
            
            cell?.setInputCountTFMaxValue(with: kDuplicateForgetTimeMaxValue)
            cell?.titleOffset = 20.0
            cell?.defaultValue = duplicateForgetTime
            cell?.updateUI(with: titleTag, valueNum: duplicateForgetTime)
            return cell!
        } else if titleTag == GeneralSettings.shared.barcodeSettings.minimumDecodeIntervalStr {
            
            let identifier = "BasicTextTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicTextTableViewCell
            if cell == nil {
                cell = BasicTextTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
            
            cell?.questionCompletion = {
                [unowned self] in
                self.handleDBRExplain(with: indexPath, settingString: titleTag)
            }
            
            cell?.inputTFValueChangedCompletion = {
                [unowned self] value in
                self.handleInputValueChanged(with: titleTag, numValue: value)
            }
            
            cell?.setInputCountTFMaxValue(with: kMinimumDecodeIntervalMaxValue)
            cell?.defaultValue = minimumDecodeInterval
            cell?.updateUI(with: titleTag, valueNum: minimumDecodeInterval)
            return cell!
        } else if titleTag == GeneralSettings.shared.barcodeSettings.continuousScanStr ||
                  titleTag == GeneralSettings.shared.barcodeSettings.resultVerificationStr ||
                  titleTag == GeneralSettings.shared.barcodeSettings.duplicateFliterStr ||
                  titleTag == GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr {
            
            let identifier = "BasicSwitchTableViewCell"
            var cell = tableView.dequeueReusableCell(withIdentifier: identifier) as? BasicSwitchTableViewCell
            if cell == nil {
                cell = BasicSwitchTableViewCell.init(style: .default, reuseIdentifier: identifier)
            }
            
            if titleTag == GeneralSettings.shared.barcodeSettings.resultVerificationStr ||
               titleTag == GeneralSettings.shared.barcodeSettings.duplicateFliterStr {
                cell?.questionButton.isHidden = false
            } else {
                cell?.questionButton.isHidden = true
            }
            
            cell?.questionCompletion = {
                [unowned self] in
                self.handleDBRExplain(with: indexPath, settingString: titleTag)
            }
            
            cell?.switchChangedCompletion = {
                [unowned self] isOn in
                self.handleDBRSettingSwitch(with: indexPath, settingString: titleTag, isOn: isOn)
            }
            
            cell?.updateUI(with: titleTag, isOn:recordDBRSwitchStateDic[titleTag]!)
            return cell!
        }
        
       
        return UITableViewCell.init()
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let titleTag = self.dbrSettingsDataArray[indexPath.row]
        
        if titleTag == GeneralSettings.shared.barcodeSettings.barcodeFormatStr {
            let barcodeFormatDetailVC = BarcodeFormatDetailViewController()
            self.navigationController?.pushViewController(barcodeFormatDetailVC, animated: true)
        }
    }
    
    private func handleDBRSettingSwitch(with indexPath: IndexPath,
                                        settingString: String,
                                        isOn: Bool) -> Void {
        if settingString == GeneralSettings.shared.barcodeSettings.continuousScanStr {
            
            GeneralSettings.shared.barcodeSettings.continuousScanIsOpen = isOn
        } else if settingString == GeneralSettings.shared.barcodeSettings.resultVerificationStr {
            
            GeneralSettings.shared.barcodeSettings.resultVerificationIsOpen = isOn
            GeneralSettings.shared.barcodeReader.enableResultVerification = isOn
        } else if settingString == GeneralSettings.shared.barcodeSettings.duplicateFliterStr {
            
            GeneralSettings.shared.barcodeSettings.duplicateFliterIsOpen = isOn
            GeneralSettings.shared.barcodeReader.enableDuplicateFilter = isOn
        } else if settingString == GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesStr {
            
            GeneralSettings.shared.barcodeSettings.decodeInvertedBarcodesIsOpen = isOn
            var grayscaleTransformationModes: [Int] = []
            if (isOn == true) {
                grayscaleTransformationModes = [EnumGrayscaleTransformationMode.original.rawValue,
                                                EnumGrayscaleTransformationMode.inverted.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue
                ]
            } else {
                grayscaleTransformationModes = [EnumGrayscaleTransformationMode.original.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue,
                                                EnumGrayscaleTransformationMode.skip.rawValue
                ]
            }
            GeneralSettings.shared.ipublicRuntimeSettings.furtherModes.grayscaleTransformationModes = grayscaleTransformationModes
            let _ = GeneralSettings.shared.updateIpublicRuntimeSettings()
        }
        
        self.handleData()
        self.dbrSettingsTableView.reloadData()
    }
    
    func handleInputValueChanged(with settingString: String,
                                 numValue: Int) -> Void {
        if settingString == GeneralSettings.shared.barcodeSettings.expectedCountStr {
            
            GeneralSettings.shared.barcodeSettings.expectedCount = numValue
            GeneralSettings.shared.ipublicRuntimeSettings.expectedBarcodesCount = numValue
            let _ = GeneralSettings.shared.updateIpublicRuntimeSettings()
        } else if settingString == GeneralSettings.shared.barcodeSettings.minimumResultConfidenceStr {
            
            GeneralSettings.shared.barcodeSettings.minimumResultConfidence = numValue
            GeneralSettings.shared.ipublicRuntimeSettings.minResultConfidence = numValue
            let _ = GeneralSettings.shared.updateIpublicRuntimeSettings()
        } else if settingString == GeneralSettings.shared.barcodeSettings.duplicateForgetTimeStr {
            
            GeneralSettings.shared.barcodeSettings.duplicateForgetTime = numValue
            GeneralSettings.shared.barcodeReader.duplicateForgetTime = numValue
        } else if settingString == GeneralSettings.shared.barcodeSettings.minimumDecodeIntervalStr {
            
            GeneralSettings.shared.barcodeSettings.minimumDecodeInterval = numValue
            GeneralSettings.shared.barcodeReader.minImageReadingInterval = numValue
        }
        
        self.handleData()
        self.dbrSettingsTableView.reloadData()
    }
    
    /// Parameter explain.
    private func handleDBRExplain(with indexPath: IndexPath,
                                  settingString: String) -> Void {
        if settingString == GeneralSettings.shared.barcodeSettings.expectedCountStr {
            ToolsManager.shared.addAlertView(to: self, title: settingString, content: expectedCountExplain, completion: nil)
        } else if settingString == GeneralSettings.shared.barcodeSettings.duplicateForgetTimeStr {
            ToolsManager.shared.addAlertView(to: self, title: settingString, content: duplicateForgetTimeExplain, completion: nil)
        } else if settingString == GeneralSettings.shared.barcodeSettings.minimumDecodeIntervalStr {
            ToolsManager.shared.addAlertView(to: self, title: settingString, content: minimumDecodeIntervalExplain, completion: nil)
        } else if settingString == GeneralSettings.shared.barcodeSettings.resultVerificationStr {
            ToolsManager.shared.addAlertView(to: self, title: settingString, content: resultVerificationExplain, completion: nil)
        } else if settingString == GeneralSettings.shared.barcodeSettings.duplicateFliterStr {
            ToolsManager.shared.addAlertView(to: self, title: settingString, content: duplicateFilterExplain, completion: nil)
        }
    }
}
